File: app/schemas/virtual_claims.py
"""Virtual claim checklist schemas."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

ChecklistStatus = Literal["missing", "present", "derived", "needs_review"]
ChecklistSource = Literal["database", "user", "llm_extracted", "derived", "policy"]


class VirtualClaimFieldInput(BaseModel):
    key: str = Field(..., min_length=1)
    value: Any = None


class VirtualClaimBootstrapRequest(BaseModel):
    patient_id: int | None = None
    insurance_company_id: int | None = None
    procedure_code: str | None = None


class VirtualClaimPatchRequest(BaseModel):
    patient_id: int | None = None
    insurance_company_id: int | None = None
    procedure_code: str | None = None
    fields: list[VirtualClaimFieldInput] = Field(default_factory=list)
    source_type: Literal["user", "llm_extracted"] = "user"


class VirtualClaimFieldResponse(BaseModel):
    key: str
    label: str
    status: ChecklistStatus
    value: Any = None
    source_type: ChecklistSource


class VirtualClaimMissingFieldResponse(BaseModel):
    key: str
    label: str
    question: str


class VirtualClaimQuestionResponse(BaseModel):
    question_key: str
    prompt: str
    status: Literal["open", "answered", "dismissed"]
    answer: Any = None


class VirtualClaimPartySummary(BaseModel):
    id: int | None = None
    name: str | None = None
    date_of_birth: str | None = None


class VirtualClaimProcedureSummary(BaseModel):
    code: str | None = None
    description: str | None = None


class VirtualClaimPolicySummaryResponse(BaseModel):
    policy_link_id: int | None = None
    policy_rule_id: int | None = None
    policy_url: str | None = None
    title: str | None = None
    extracted_at: datetime | None = None
    rules_json: Any = None
    criteria_json: Any = None
    notes_json: Any = None


class VirtualClaimChecklistValue(BaseModel):
    value: Any = None
    status: ChecklistStatus
    source_type: ChecklistSource
    required: bool = True
    label: str | None = None


class VirtualClaimPatientChecklist(BaseModel):
    patient_id: VirtualClaimChecklistValue
    first_name: VirtualClaimChecklistValue
    last_name: VirtualClaimChecklistValue
    date_of_birth: VirtualClaimChecklistValue


class VirtualClaimPayerChecklist(BaseModel):
    insurance_company_id: VirtualClaimChecklistValue
    payer_name: VirtualClaimChecklistValue
    member_id: VirtualClaimChecklistValue
    group_number: VirtualClaimChecklistValue
    policy_number: VirtualClaimChecklistValue


class VirtualClaimServiceChecklist(BaseModel):
    procedure_code: VirtualClaimChecklistValue
    procedure_description: VirtualClaimChecklistValue
    service_date: VirtualClaimChecklistValue
    rendering_provider: VirtualClaimChecklistValue
    quantity: VirtualClaimChecklistValue
    modifier: VirtualClaimChecklistValue


class VirtualClaimDiagnosisChecklist(BaseModel):
    diagnosis_code: VirtualClaimChecklistValue
    diagnosis_description: VirtualClaimChecklistValue


class VirtualClaimPolicyChecklist(BaseModel):
    policy_link_id: VirtualClaimChecklistValue
    policy_url: VirtualClaimChecklistValue
    stored_rules_available: VirtualClaimChecklistValue
    radiculopathy_evidence: VirtualClaimChecklistValue
    dermatomal_distribution: VirtualClaimChecklistValue
    functional_limitation: VirtualClaimChecklistValue
    conservative_treatment_failed: VirtualClaimChecklistValue
    imaging_guidance: VirtualClaimChecklistValue
    MRI_or_CT_or_EMG_evidence: VirtualClaimChecklistValue
    neuro_exam_evidence: VirtualClaimChecklistValue
    frequency_session_limits_respected: VirtualClaimChecklistValue
    radiologic_findings_consistent: VirtualClaimChecklistValue | None = None
    initial_therapeutic_tfesi: VirtualClaimChecklistValue | None = None
    vertebral_level_limits_respected: VirtualClaimChecklistValue | None = None


class VirtualClaimReadinessChecklist(BaseModel):
    ready_to_draft: bool
    missing_fields: list[str] = Field(default_factory=list)
    blocking_reasons: list[str] = Field(default_factory=list)
    next_questions: list[str] = Field(default_factory=list)


class VirtualClaimChecklistResponse(BaseModel):
    patient: VirtualClaimPatientChecklist
    payer_insurance: VirtualClaimPayerChecklist
    service: VirtualClaimServiceChecklist
    diagnosis: VirtualClaimDiagnosisChecklist
    policy_medical_necessity: VirtualClaimPolicyChecklist
    readiness: VirtualClaimReadinessChecklist


class VirtualClaimResponse(BaseModel):
    draft_id: int
    session_id: int
    status: Literal["open", "ready", "materialized", "archived"]
    readiness: bool
    readiness_reason: str | None = None
    patient: VirtualClaimPartySummary | None = None
    payer: VirtualClaimPartySummary | None = None
    procedure: VirtualClaimProcedureSummary | None = None
    materialized_claim_id: int | None = None
    policy_summary: VirtualClaimPolicySummaryResponse | None = None
    checklist: VirtualClaimChecklistResponse
    filled: list[VirtualClaimFieldResponse] = Field(default_factory=list)
    missing: list[VirtualClaimFieldResponse] = Field(default_factory=list)
    needs_review: list[VirtualClaimFieldResponse] = Field(default_factory=list)
    policy_constraints: list[VirtualClaimFieldResponse] = Field(default_factory=list)
    missing_fields: list[VirtualClaimMissingFieldResponse] = Field(default_factory=list)
    follow_up_questions: list[VirtualClaimQuestionResponse] = Field(default_factory=list)
    updated_at: datetime | None = None


class VirtualClaimMaterializeRequest(BaseModel):
    confirm: bool = False


class VirtualClaimMaterializeResponse(BaseModel):
    action_required: bool = False
    proposal: dict[str, Any] | None = None
    claim_id: int | None = None
    draft: VirtualClaimResponse

File: frontend-dev/src/api/virtualClaims.ts
import { z } from "zod";

import { requestJson } from "@/lib/api/client";
import type { VirtualClaimDTO, VirtualClaimMaterializeDTO } from "@/types/virtualClaim";

const virtualClaimFieldSchema = z.object({
  key: z.string(),
  label: z.string(),
  status: z.enum(["missing", "present", "derived", "needs_review"]),
  value: z.unknown().nullable().optional(),
  source_type: z.enum(["database", "user", "llm_extracted", "derived", "policy"]),
});

const virtualClaimMissingFieldSchema = z.object({
  key: z.string(),
  label: z.string(),
  question: z.string(),
});

const virtualClaimQuestionSchema = z.object({
  question_key: z.string(),
  prompt: z.string(),
  status: z.enum(["open", "answered", "dismissed"]),
  answer: z.unknown().nullable().optional(),
});

const virtualClaimPartySchema = z.object({
  id: z.number().nullable().optional(),
  name: z.string().nullable().optional(),
  date_of_birth: z.string().nullable().optional(),
});

const virtualClaimProcedureSchema = z.object({
  code: z.string().nullable().optional(),
  description: z.string().nullable().optional(),
});

const virtualClaimPolicySummarySchema = z.object({
  policy_link_id: z.number().nullable().optional(),
  policy_rule_id: z.number().nullable().optional(),
  policy_url: z.string().nullable().optional(),
  title: z.string().nullable().optional(),
  extracted_at: z.string().nullable().optional(),
  rules_json: z.unknown().nullable().optional(),
  criteria_json: z.unknown().nullable().optional(),
  notes_json: z.unknown().nullable().optional(),
});

const virtualClaimSchema = z.object({
  draft_id: z.number(),
  session_id: z.number(),
  status: z.enum(["open", "ready", "materialized", "archived"]),
  readiness: z.boolean(),
  readiness_reason: z.string().nullable().optional(),
  patient: virtualClaimPartySchema.nullable().optional(),
  payer: virtualClaimPartySchema.nullable().optional(),
  procedure: virtualClaimProcedureSchema.nullable().optional(),
  materialized_claim_id: z.number().nullable().optional(),
  policy_summary: virtualClaimPolicySummarySchema.nullable().optional(),
  filled: z.array(virtualClaimFieldSchema),
  missing: z.array(virtualClaimFieldSchema),
  needs_review: z.array(virtualClaimFieldSchema),
  policy_constraints: z.array(virtualClaimFieldSchema),
  missing_fields: z.array(virtualClaimMissingFieldSchema),
  follow_up_questions: z.array(virtualClaimQuestionSchema),
  updated_at: z.string().nullable().optional(),
});

const virtualClaimMaterializeSchema = z.object({
  action_required: z.boolean(),
  proposal: z.record(z.string(), z.unknown()).nullable().optional(),
  claim_id: z.number().nullable().optional(),
  draft: virtualClaimSchema,
});

function parseWithSchema<T>(schema: z.ZodType<T>, data: unknown, label: string): T {
  const parsed = schema.safeParse(data);
  if (!parsed.success) {
    const issues = parsed.error.issues.map((issue) => issue.message).join("; ");
    throw new Error(`Invalid ${label} response: ${issues}`);
  }
  return parsed.data;
}

export async function ensureVirtualClaim(
  sessionId: number,
  payload: {
    patient_id?: number;
    insurance_company_id?: number;
    procedure_code?: string;
  } = {}
): Promise<VirtualClaimDTO> {
  const data = await requestJson<unknown>(`/api/chat/sessions/${sessionId}/virtual-claim`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return parseWithSchema(virtualClaimSchema, data, "ensure virtual claim");
}

export async function getVirtualClaim(sessionId: number): Promise<VirtualClaimDTO> {
  const data = await requestJson<unknown>(`/api/chat/sessions/${sessionId}/virtual-claim`);
  return parseWithSchema(virtualClaimSchema, data, "get virtual claim");
}

export async function patchVirtualClaim(
  sessionId: number,
  payload: {
    patient_id?: number;
    insurance_company_id?: number;
    procedure_code?: string;
    source_type?: "user" | "llm_extracted";
    fields?: Array<{ key: string; value: unknown }>;
  }
): Promise<VirtualClaimDTO> {
  const data = await requestJson<unknown>(`/api/chat/sessions/${sessionId}/virtual-claim`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return parseWithSchema(virtualClaimSchema, data, "patch virtual claim");
}

export async function recomputeVirtualClaim(sessionId: number): Promise<VirtualClaimDTO> {
  const data = await requestJson<unknown>(
    `/api/chat/sessions/${sessionId}/virtual-claim/recompute`,
    {
      method: "POST",
    }
  );
  return parseWithSchema(virtualClaimSchema, data, "recompute virtual claim");
}

export async function materializeVirtualClaim(
  sessionId: number,
  confirm = false
): Promise<VirtualClaimMaterializeDTO> {
  const data = await requestJson<unknown>(
    `/api/chat/sessions/${sessionId}/virtual-claim/materialize`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ confirm }),
    }
  );
  return parseWithSchema(
    virtualClaimMaterializeSchema,
    data,
    "materialize virtual claim"
  );
}
